src/proto/smb.rs

```diff
@@ -0,0 +1,1150 @@
+        let time: u64 = (EPOCH_1601 + chrono::Utc::now().timestamp() as u64) * (1e7 as u64);
+        // Chose dialect
+        let smb2_versions = [
+            (0x0202, "SMB 2.002"),
+            (0x0210, "SMB 2.1"),
+            (0x02ff, "SMB 2.???"),
+            (0x0300, "SMB 3.0"),
+            (0x0302, "SMB 2.0.2"),
+            (0x0310, "SMB 3.1.0"),
+            (0x0311, "SMB 3.1.1"),
+        ];
+        let mut dialect = None;
+        let mut dialect_name = "Unknown";
+        if let Some(smb_ver) = smb2_versions
+            .iter()
+            .find(|(d, _)| self.dialects.contains(d))
+        {
+            dialect = Some(smb_ver.0);
+            dialect_name = smb_ver.1;
+        }
+        resp.extend_from_slice(&0x41_u16.to_le_bytes()); // StructureSize
+        resp.extend_from_slice(&0x1_u16.to_le_bytes()); // SecurityMode
+        resp.extend_from_slice(&dialect?.to_le_bytes()); // DialectRevision
+        resp.extend_from_slice(&0x1_u16.to_le_bytes()); // NegotiateCount
+        resp.extend_from_slice(&self.client_guid); // GUID
+        resp.extend_from_slice(&0x1_u32.to_le_bytes()); // Capabilities
+        resp.extend_from_slice(&0x10000_u32.to_le_bytes()); // MaxTransactionSize
+        resp.extend_from_slice(&0x10000_u32.to_le_bytes()); // MaxReadSize
+        resp.extend_from_slice(&0x10000_u32.to_le_bytes()); // MaxWriteSize
+        resp.extend_from_slice(&time.to_le_bytes()); // ServerTime
+        resp.extend_from_slice(&time.to_le_bytes()); // ServerStartTime
+        resp.extend_from_slice(&0x80_u16.to_le_bytes()); // SecurityBloboffset
+        resp.extend_from_slice(&(SECURITY_BLOB.len() as u16).to_le_bytes()); // SecurityBlobLength
+        resp.extend_from_slice(&0x0_u32.to_le_bytes()); // NegotiateContextOffset
+        resp.extend_from_slice(SECURITY_BLOB); // SecurityBlob
+        warn!("SMB2 Negotiate-Protocol-Reply ({})", dialect_name);
+        Some(resp)
+    }
+}
+
+// #[derive(Debug, Clone)]
+// struct SMB2SetupRequest {}
+// impl SMB2SetupRequest {
+//     fn new() -> Self {
+//         SMB2SetupRequest {}
+//     }
+
+//     fn repl(&self) -> Option<Vec<u8>> {
+//         (None, None)
+//     }
+// }
+
+#[derive(Debug, Clone)]
+enum SMB2Payload {
+    NegotiateRequest(SMB2NegotiateRequest),
+    // SetupRequest(SMB2SetupRequest),
+}
+
+impl SMB2Payload {
+    fn repl(&self) -> Option<Vec<u8>> {
+        match self {
+            SMB2Payload::NegotiateRequest(x) => x.repl(),
+            // SMB2Payload::SetupRequest(x) => x.repl(),
+        }
+    }
+    fn parse(&mut self, byte: &u8) {
+        match self {
+            SMB2Payload::NegotiateRequest(x) => x.parse(byte),
+            // SMB2Payload::SetupRequest(x) => x.repl(),
+        }
+    }
+}
+
+//////////////
+// Handlers //
+//////////////
+
+pub fn repl_smb1<'a>(
+    data: &'a [u8],
+    _masscanned: &Masscanned,
+    _client_info: &ClientInfo,
+) -> Option<Vec<u8>> {
+    let mut nbtsession: NBTSession<SMB1Header> = NBTSession::new();
```

Comment:
Note: this is re-created on each call to `repl_smb1` which is not expected: we should store it somewhere in `client_info`. Waiting for PR #21 as discussed with frk